File: binding/src/libqmimetype.cpp
#include <QMimeType>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qmimetype.h>
#include "libqmimetype.h"
#include "libqmimetype.hxx"

QMimeType* QMimeType_new() {
	 return new QMimeType();
}

QMimeType* QMimeType_new2(const QMimeType* other) {
	 return new QMimeType(*other);
}

QMimeType* QMimeType_new3(const QMimeTypePrivate* dd) {
	 return new QMimeType(*dd);
}

void QMimeType_OperatorAssign(QMimeType* self, const QMimeType* other) {
	self->operator=(*other);
}

void QMimeType_Swap(QMimeType* self, QMimeType* other) {
	self->swap(*other);
}

bool QMimeType_IsValid(const QMimeType* self) {
	return self->isValid();
}

bool QMimeType_IsDefault(const QMimeType* self) {
	return self->isDefault();
}

libqt_string QMimeType_Name(const QMimeType* self) {
	QString _ret = self->name();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMimeType_Comment(const QMimeType* self) {
	QString _ret = self->comment();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMimeType_GenericIconName(const QMimeType* self) {
	QString _ret = self->genericIconName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QMimeType_IconName(const QMimeType* self) {
	QString _ret = self->iconName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QMimeType_GlobPatterns(const QMimeType* self) {
	return self->globPatterns();
}

libqt_list QMimeType_ParentMimeTypes(const QMimeType* self) {
	return self->parentMimeTypes();
}

libqt_list QMimeType_AllAncestors(const QMimeType* self) {
	return self->allAncestors();
}

libqt_list QMimeType_Aliases(const QMimeType* self) {
	return self->aliases();
}

libqt_list QMimeType_Suffixes(const QMimeType* self) {
	return self->suffixes();
}

libqt_string QMimeType_PreferredSuffix(const QMimeType* self) {
	QString _ret = self->preferredSuffix();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QMimeType_Inherits(const QMimeType* self, const libqt_string mimeTypeName) {
	return self->inherits(QString::fromUtf8(mimeTypeName.data, mimeTypeName.len));
}

libqt_string QMimeType_FilterString(const QMimeType* self) {
	QString _ret = self->filterString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QMimeType_Delete(QMimeType* self) {
    delete self;
}

File: binding/include/libqmimetype.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmimetype.html
pub const qmimetype = struct {

    /// New constructs a new QMimeType object.
    pub fn New() QtC.QMimeType {
        return qtc.QMimeType_new();
    }


    /// New2 constructs a new QMimeType object.
    pub fn New2(other: ?*anyopaque) QtC.QMimeType {
        return qtc.QMimeType_new2(@ptrCast(other));
    }


    /// New3 constructs a new QMimeType object.
    pub fn New3(dd: *const QMimeTypePrivate) QtC.QMimeType {
        return qtc.QMimeType_new3(@ptrCast(dd));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMimeType_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMimeType_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QMimeType_IsValid(@ptrCast(self));
    }

    pub fn IsDefault(self: ?*anyopaque, ) bool {
        return qtc.QMimeType_IsDefault(@ptrCast(self));
    }

    pub fn Name(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeType_Name(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimetype.Name: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Comment(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeType_Comment(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimetype.Comment: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn GenericIconName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeType_GenericIconName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimetype.GenericIconName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IconName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeType_IconName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimetype.IconName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn GlobPatterns(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeType_GlobPatterns(@ptrCast(self));
    }

    pub fn ParentMimeTypes(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeType_ParentMimeTypes(@ptrCast(self));
    }

    pub fn AllAncestors(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeType_AllAncestors(@ptrCast(self));
    }

    pub fn Aliases(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeType_Aliases(@ptrCast(self));
    }

    pub fn Suffixes(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeType_Suffixes(@ptrCast(self));
    }

    pub fn PreferredSuffix(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeType_PreferredSuffix(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimetype.PreferredSuffix: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Inherits(self: ?*anyopaque, mimeTypeName: []const u8) bool {
        const mimeTypeName_str = qtc.libqt_string{
    .len = mimeTypeName.len,
    .data = mimeTypeName.ptr,
};
return qtc.QMimeType_Inherits(@ptrCast(self), mimeTypeName_str);
    }

    pub fn FilterString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QMimeType_FilterString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimetype.FilterString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMimeType_Delete(@ptrCast(self));
    }
};
